loopy/codegen/tools.py

```diff
@@ -0,0 +1,152 @@
+                            Barrier, BeginBlockItem, gather_schedule_block)
+
+
+__doc__ = """
+.. currentmodule:: loopy.codegen.tools
+
+.. autoclass:: CodegenOperationCacheManager
+"""
+
+
+class CodegenOperationCacheManager(ImmutableRecord):
+    """
+    Caches operations arising during the codegen pipeline.
+    """
+    def __init__(self, kernel):
+        super().__init__()
+        super().__setattr__("kernel", kernel)
+
+    def __setattr__(self, key, val):
+        if key == "kernel":
+            # overwriting self.kernel is not allowed, as cache will become
+            # invalidated
+            raise ValueError("Cannot update CodegenOperationCacheManager.kernel,"
+                             " create a new instance instead.")
+
+        super().__setattr__(key, val)
+
+    @memoize_method
+    def find_active_inames_at(self, sched_index):
+        """
+        Returns a :class:`frozenset` of active inames at the point just before
+        *sched_index*.
+        """
+        if sched_index == 0:
+            return frozenset()
+
+        sched_item = self.kernel.schedule[sched_index-1]
+        if isinstance(sched_item, EnterLoop):
+            return (self.find_active_inames_at(sched_index-1)
+                    | frozenset([sched_item.iname]))
+        elif isinstance(sched_item, LeaveLoop):
+            assert sched_item.iname in self.find_active_inames_at(sched_index-1)
+            return (self.find_active_inames_at(sched_index-1)
+                    - frozenset([sched_item.iname]))
+        else:
+            return self.find_active_inames_at(sched_index-1)
+
+    @memoize_method
+    def has_barrier_within(self, sched_index):
+        """
+        Checks if the kernel's schedule block from *sched_index* contains a
+        barrier.
+        """
+        sched_item = self.kernel.schedule[sched_index]
+
+        if isinstance(sched_item, BeginBlockItem):
+            # TODO: calls to "gather_schedule_block" can be amortized
+            _, endblock_index = gather_schedule_block(self.kernel.schedule,
+                    sched_index)
+            return any(self.has_barrier_within(i)
+                       for i in range(sched_index+1, endblock_index))
+        elif isinstance(sched_item, Barrier):
+            return True
+        else:
+            return False
+
+    @memoize_method
+    def get_callkernel_index(self, sched_index):
+        """
+        Returns index of :class:`loopy.schedule.CallKernel` containing the point
+        just before *sched_index*. Return *None* if the point is
+        not a part of any sub-kernel.
+        """
+        if sched_index == 0:
+            return None
+
+        sched_item = self.kernel.schedule[sched_index-1]
+
+        if isinstance(sched_item, CallKernel):
+            return sched_index-1
+        elif isinstance(sched_item, ReturnFromKernel):
+            return None
+        else:
+            return self.get_callkernel_index(sched_index-1)
```

Comment:
While/for loop instead of recursion? Otherwise you might run into recursion limits. Also, Python (sadly) doesn't do tail calls optimization.

Reply:
Should be fixed now